accounts-db/src/accounts_hash.rs

```diff
@@ -1064,22 +1065,61 @@ impl<'a> AccountsHasher<'a> {
                 }
             }
 
-            if !duplicate_pubkey_indexes.is_empty() {
-                // skip past duplicate keys in earlier slots
-                // reverse this list because get_item can remove first_items[*i] when *i is exhausted
-                //  and that would mess up subsequent *i values
-                duplicate_pubkey_indexes.iter().rev().for_each(|i| {
-                    Self::get_item(
-                        *i,
-                        pubkey_bin,
-                        &mut first_items,
-                        sorted_data_by_pubkey,
-                        &mut indexes,
-                        &mut first_item_to_pubkey_division,
-                        &binner,
-                    );
+            // looping to add next item to working set
+            while let Some((key, slot_group_index, offset)) = next {
+                // if `new key` is less than the min key in the working set, continue process with `new key` directly.
+                if let Some((current_min_slot_group_index, current_min_offset)) = working_set.last()
+                {
+                    let current_min_key = &sorted_data_by_pubkey[*current_min_slot_group_index]
+                        [*current_min_offset]
+                        .pubkey;
+                    if key < current_min_key {
+                        new = Some((slot_group_index, offset));
+                        break;
+                    }
+                }
+
+                let found = working_set.binary_search_by(|(slot_group_index, offset)| {
+                    let prob = &sorted_data_by_pubkey[*slot_group_index][*offset].pubkey;
+                    key.cmp(prob)
                 });
-                duplicate_pubkey_indexes.clear();
+
+                match found {
+                    Err(index) => {
+                        // found a new new key, insert into the working_set
+                        working_set.insert(index, (slot_group_index, offset));
+                        break;
+                    }
+                    Ok(index) => {
+                        let v = &mut working_set[index];
+                        if v.0 > slot_group_index {
+                            // There is already a later slot group that contains this key in the working_set,
+                            // look up again.
+                            let (_item, new_next) = Self::get_item(
```

Comment:
maybe both of these get_item calls:
`next = Self::get_item(...).1;`
Or, if you make a type out of `get_item`'s return value:
`next = Self::get_item(...).next;`